File: src/leetcode/hash_maps_and_sets/2352_equal_row_and_column_pairs.rb
# frozen_string_literal: true

# LeetCode 75
# https://leetcode.com/problems/equal-row-and-column-pairs/description/?envType=study-plan-v2&envId=leetcode-75

# Time complexity: O(n^2)
# Space complexity: O(n^2)

# @param {Integer[][]} grid
# @return {Integer}
def equal_pairs(grid)
  rows = grid.each_with_object({}).with_index(1) do |(row, hash_map), index|
    hash_map[index] = row
  end

  columns = grid.transpose.each_with_object({}).with_index(1) do |(column, hash_map), index|
    hash_map[index] = column
  end

  equal_pairs_count = 0

  rows.each_value do |row|
    columns.each_value do |column|
      equal_pairs_count += 1 if row == column
    end
  end

  equal_pairs_count
end

def equal_pairs_clean(grid)
  count = 0

  rows = grid.tally

  grid.transpose.each { |col| count += rows[col] if rows[col] }

  count
end

def equal_pairs_optimized(grid)
  count = 0

  row_counts = grid.tally
  # row_counts = Hash.new { |hash, key| hash[key] = 0 }
  # grid.each { |row| row_counts[row] += 1 }

  n = grid.length
  (0...n).each do |col|
    column = Array.new(n) { |row| grid[row][col] }

    count += row_counts[column] if row_counts[column]
  end

  count
end

File: src/leetcode/hash_maps_and_sets/2352_equal_row_and_column_pairs.ts
// LeetCode 75
// https://leetcode.com/problems/equal-row-and-column-pairs/description/?envType=study-plan-v2&envId=leetcode-75

// Time complexity: O(n^2)
// Space complexity: O(n ^ 2)

const equalPairs = (grid: number[][]): number => {
  let count = 0;

  const rowCounts = grid.reduce((acc, row) => {
    const key = row.join(',');

    acc.set(key, (acc.get(key) ?? 0) + 1);

    return acc;
  }, new Map<string, number>());

  const n = grid.length;

  for (let col = 0; col < n; col++) {
    const column: number[] = [];

    for (let row = 0; row < n; row++) {
      column.push(grid[row][col]);
    }

    const key = column.join(',');

    if (rowCounts.has(key)) {
      count += rowCounts.get(key) || 0;
    }
  }

  return count;
};

export default equalPairs;
